tracing-appender/src/rolling.rs

```diff
@@ -0,0 +1,178 @@
+use crate::inner::InnerAppender;
+use chrono::{DateTime, Datelike, TimeZone, Timelike, Utc};
+use std::fmt::Debug;
+use std::fs::{File, OpenOptions};
+use std::io::{BufWriter, Write};
+use std::path::Path;
+use std::sync::{Mutex, MutexGuard};
+use std::{fs, io};
+use tracing_subscriber::fmt::MakeWriter;
+
+pub struct RollingFileAppender {
+    inner: Mutex<InnerAppender<BufWriterFactory>>,
+}
+
+impl RollingFileAppender {
+    fn writer(&self) -> RollingFileWriter {
+        RollingFileWriter::new(&self.inner)
+    }
+}
+
+pub struct RollingFileWriter<'a> {
+    inner: &'a Mutex<InnerAppender<BufWriterFactory>>,
+}
+
+impl<'a> RollingFileWriter<'a> {
+    fn new(inner: &'a Mutex<InnerAppender<BufWriterFactory>>) -> Self {
+        Self { inner }
+    }
+
+    fn inner(&self) -> io::Result<MutexGuard<'a, InnerAppender<BufWriterFactory>>> {
+        self.inner.lock().map_err(
+            |_: std::sync::PoisonError<MutexGuard<InnerAppender<BufWriterFactory>>>| {
+                io::Error::from(io::ErrorKind::Other)
+            },
+        )
+    }
+}
+
+impl<'a> io::Write for RollingFileWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner()?.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner()?.flush()
+    }
+}
+
+impl<'a> MakeWriter for RollingFileWriter<'a> {
+    type Writer = RollingFileWriter<'a>;
+
+    fn make_writer(&self) -> Self::Writer {
+        RollingFileWriter::new(&self.inner)
+    }
+}
+
+pub fn hourly<P: AsRef<Path>>(directory: P, file_name_prefix: P) -> RollingFileAppender {
+    create_writer(
+        Rotation::HOURLY,
+        directory.as_ref(),
+        file_name_prefix.as_ref(),
+    )
+}
+
+pub fn daily<P: AsRef<Path>>(directory: P, file_name_prefix: P) -> RollingFileAppender {
+    create_writer(
+        Rotation::DAILY,
+        directory.as_ref(),
+        file_name_prefix.as_ref(),
+    )
+}
+
+pub fn never<P: AsRef<Path>>(directory: P, file_name_prefix: P) -> RollingFileAppender {
+    create_writer(
+        Rotation::NEVER,
+        directory.as_ref(),
+        file_name_prefix.as_ref(),
+    )
+}
+
+fn create_writer(
```

Comment:
I think this should probably be a public `RollingFileAppender::new` constructo taking paths and a `Rotation` policy. The helper constructors `hourly`, `weekly`, and `never` are nice, but I think they should be implemented in terms of a `new` constructor.